### ISSUE TITLE
Bug: [switch-exhaustiveness-check] fails with option "allowDefaultCaseForExhaustiveSwitch": false

### ISSUE THREAD

**Issue by danielrentz** [typescript-eslint/typescript-eslint#8165]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=GYVwdgxgLglg9mABFApgZygCjALkWEAWwCMUAnASkQG8BYAKEUTQHcYoIALRbKuxpoggBDNCkQBGPGRRQQZJACYA3A0FDR4xdNnykAFlUCmAExTBhIADZQdchYgBsRpgF8G7%2BkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6MAG1IiaNAD20SABpMUYfHi6OAEUQAzUfHwBhYckQAxfQFEho8VMQAZS4eXjV7eDcwAF8AXXAYkGigA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function test(n: number) {
  switch (n) {
    case 1: return 2;
    case 2: return 4;
    default: return 6;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/switch-exhaustiveness-check": ["error", { "allowDefaultCaseForExhaustiveSwitch": false }]
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No error.

### Actual Result

The switch statement is exhaustive, so the default case is unnecessary. 5:5 - 5:23

**Comment by ehoogeveen-medweb** [comment]
In fact, with `"requireDefaultForNonUnion": true` also set it is not possible to write a compliant switch:
- [with default](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=GYVwdgxgLglg9mABFApgZygCjALkWEAWwCMUAnASkQG8BYAKEUTQHcYoIALRbKuxpoggBDNCkQBGPGRRQQZJACYA3A0FDR4xdNnykAFlUCmAExTBhIADZQdchYgBsRpgF8G7%2BkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6MAG1w2HImjQA9tEgAaLdqzbskYfHj6OAEUQAzUfHwBhYckQAxQwCiQqLiUogAylw8vGqu8D6mllbQiACOsJQpTq4I%2BP7QAHL6TACqTJTFavhwiGbYAL51ALpajfVAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) gives "The switch statement is exhaustive, so the default case is unnecessary. 5:5 - 5:23"
- [no default](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=GYVwdgxgLglg9mABFApgZygCjALkWEAWwCMUAnASkQG8BYAKEUTQHcYoIALRbKuxpoggBDNCkQBGPGRRQQZJACYA3A0FDR4xdNnykAFlUCAvg1P0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6MAG1w2HImjQA9tEgAaLdqzbskYfHj6OAEUQAzUfHwBhYckQAxQwCiQqLiUogAylw8vGqu8D6mllbQiACOsJQpTq4I%2BP7QAHL6TACqTJTFavhwiGbYAL51ALpajfVAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) gives "Switch is not exhaustive. Cases not matched: default 2:11 - 2:12"

I wonder if the problem is that the switch statement is exhaustive *only because* of the default - so its presence makes it redundant, and its absence makes it necessary.

**Comment by Josh-Cena** [comment]
An infinite type definitely needs a default. I'm surprised this isn't covered by a test—cc @Zamiell

**Comment by Zamiell** [comment]
That looks like a bug, I'll investigate.

**Comment by Zamiell** [comment]
should be fixed in this PR, sorry about that

**Comment by danielrentz** [comment]
Thank you for the quick fix!
